src/agent_observer.py

```diff
@@ -0,0 +1,104 @@
+
+        Args:
+            charm: The parent charm to attach the observer to.
+            state: The charm state.
+            jenkins_agent_service: Service manager that controls Jenkins agent service.
+        """
+        super().__init__(charm, "agent-observer")
+        self.charm = charm
+        self.state = state
+        self.jenkins_agent_service = jenkins_agent_service
+        charm.framework.observe(
+            charm.on[AGENT_RELATION].relation_joined, self._on_agent_relation_joined
+        )
+        charm.framework.observe(
+            charm.on[AGENT_RELATION].relation_changed, self._on_agent_relation_changed
+        )
+        charm.framework.observe(
+            charm.on[AGENT_RELATION].relation_departed, self._on_agent_relation_departed
+        )
+
+    def _on_agent_relation_joined(self, event: ops.RelationJoinedEvent) -> None:
+        """Handle agent relation joined event.
+
+        Args:
+            event: The event fired when an agent has joined the relation.
+        """
+        self.charm.unit.status = ops.MaintenanceStatus(
+            f"Setting up '{event.relation.name}' relation."
+        )
+
+        relation_data = self.state.agent_meta.get_jenkins_agent_v0_interface_dict()
+        logger.debug("Agent relation data set: %s", relation_data)
+        event.relation.data[self.charm.unit].update(relation_data)
+
+    def _on_agent_relation_changed(self, event: ops.RelationChangedEvent) -> None:
+        """Handle agent relation changed event.
+
+        Args:
+            event: The event fired when the agent relation data has changed.
+        """
+        logger.info("%s relation changed.", event.relation.name)
+
+        # Check if the pebble service has started and set agent ready.
+        if os.path.exists(str(AGENT_READY_PATH)) and self.jenkins_agent_service.is_active:
+            logger.warning("Given agent already registered. Skipping.")
+            return
+
+        if not self.state.agent_relation_credentials:
+            self.charm.unit.status = ops.WaitingStatus("Waiting for complete relation data.")
+            logger.info("Waiting for complete relation data.")
+            return
+
+        self.charm.unit.status = ops.MaintenanceStatus("Starting jenkins agent service")
+        self.jenkins_agent_service.restart()
+
+        if not self.jenkins_agent_service.is_active:
+            # The jenkins server sets credentials one by one, hence if the current credentials are
+            # not for this particular agent, the agent operator should wait until it receives one
+            # designated for it.
+            logger.warning(
+                "Failed credential for agent %s, will wait for next credentials to be set",
+                self.state.agent_meta.name,
+            )
+            self.charm.unit.status = ops.WaitingStatus(
+                "Failed to start the service. Waiting for another set of credentials"
+            )
```

Comment:
Should this be done as soon as we know that the relation data does not have the necessary credentials rather than based on whether the service is active? It is kind of implicit

Perhaps this method could be structured as, currently it was a little hard to follow:

1. Check pre-conditions, e.g., whether the necessary credentials have arrived
1. Determine whether the service is active already, if so perform any logic to determine whether it needs to be restarted handling failure cases etc
2. If the service has not started, start the service

Reply:
Agreed, I removed this part, and modify the hook to follow the logic that you outlined. I do think however that we should systematically restart the service every time in this hook in case of a jenkins version update that brought a new version of the agent binary for example.

In the future we can also implement a checksum for the agent binary which will let us potentially skip the binary download on restart.